lightning/src/onion_message/async_payments.rs

```diff
@@ -0,0 +1,142 @@
+// This file is Copyright its original authors, visible in version control
+// history.
+//
+// This file is licensed under the Apache License, Version 2.0 <LICENSE-APACHE
+// or http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your option.
+// You may not use this file except in accordance with one or both of these
+// licenses.
+
+//! Message handling for async payments.
+
+use crate::io;
+use crate::ln::msgs::DecodeError;
+use crate::onion_message::messenger::PendingOnionMessage;
+use crate::onion_message::messenger::{Responder, ResponseInstruction};
+use crate::onion_message::packet::OnionMessageContents;
+use crate::prelude::*;
+use crate::util::ser::{Readable, ReadableArgs, Writeable, Writer};
+
+// TLV record types for the `onionmsg_tlv` TLV stream as defined in BOLT 4.
+const HELD_HTLC_AVAILABLE_TLV_TYPE: u64 = 72;
+const RELEASE_HELD_HTLC_TLV_TYPE: u64 = 74;
+
+/// A handler for an [`OnionMessage`] containing an async payments message as its payload.
+///
+/// [`OnionMessage`]: crate::ln::msgs::OnionMessage
+pub trait AsyncPaymentsMessageHandler {
+	/// Handles the given message by either releasing a held HTLC, sending a payment, or replying with
+	/// an error.
+	///
+	/// The returned [`AsyncPaymentsMessage`], if any, is enqueued to be sent by [`OnionMessenger`].
+	///
+	/// [`OnionMessenger`]: crate::onion_message::messenger::OnionMessenger
+	fn handle_message(
```

Comment:
On a related note, for message authentication it may make sense to move towards separate methods if each would need different data to authenticate (e.g, handling an invoice request may need a nonce from an offer's blinded path whereas as handling an invoice may need a payment id from the invoice request's reply path). See #3085.

But we may still want an enum for the return type if a method could return different types (e.g., `Botl12Invoice` or `InvoiceError`). There may need to be different enums depending on how strict we want to make the compiler checks. Currently, we allow any `OffersMessage` variant to be returned even though some aren't really allowed.

Reply:
Updated to have per-message methods in the trait. Also added default implementations which reduced the diff a bit.